### PR TITLE
Added is_causal mask argument to flax.nnx.dot_product_attention

### PR DESCRIPTION
**Pull request by ibbyml**
# What does this PR do?

- Adds an `is_causal` arg to `flax.nnx.dot_product_attention` and `dot_product_attention_weights`.
- Forwards `is_causal` through to `jax.nn.dot_product_attention` fast path when possible.
- Implements a manual causal masking logic that:
  - Supports both self-attention and cross-attention
  - Composes `is_causal` with input masks with the `combine_masks` helper
- Adds 3 attention tests to ensure compatibility and correctness.

## Checklist

- [x] This change is discussed in this [discussion](https://github.com/google/flax/discussions/5091).
- [x] The documentation and docstrings adhere to the [documentation guidelines](https://github.com/google/flax/blob/main/docs/README.md#how-to-write-code-documentation).
- [x] This change includes necessary high-coverage tests. (No quality testing = no merge!)

### CODE REVIEW

File: flax/nnx/nn/attention.py [hunk 1]
```diff
@@ -198,56 +211,46 @@ def dot_product_attention(
...
   query, key, value = promote_dtype((query, key, value), dtype=dtype)  # type: ignore[bad-unpacking]
   dtype = query.dtype
   assert key.ndim == query.ndim == value.ndim, 'q, k, v must have same rank.'
-  assert (
-    query.shape[:-3] == key.shape[:-3] == value.shape[:-3]
-  ), 'q, k, v batch dims must match.'
-  assert (
-    query.shape[-2] == key.shape[-2] == value.shape[-2]
-  ), 'q, k, v num_heads must match.'
+  assert query.shape[:-3] == key.shape[:-3] == value.shape[:-3], 'q, k, v batch dims must match.'
```
Comment (vfdev-5) [inline flax/nnx/nn/attention.py]: Can we please revert all unrelated changes?

File: tests/nnx/nn/attention_test.py [hunk 2]
```diff
@@ -132,6 +118,70 @@ def test_keep_rngs(self, keep_rngs):
     else:
       nnx.split(module, nnx.Param)
 
+  def test_causal_mask_equivalence(self):
```
Comment (vfdev-5) [inline tests/nnx/nn/attention_test.py]: Can we make these tests parametrized?

**Comment by ibbyml** [comment]
Pushed the updated parameterized tests and removed the unnecessary formatting. The tests now cover self-attention with and without a padding mask as well as cross-attention with and without a padding mask. Happy to adjust anything else.

**Comment by vfdev-5** [comment]
> Pushed the updated parameterized tests and removed the unnecessary formatting. The tests now cover self-attention with and without a padding mask as well as cross-attention with and without a padding mask. Happy to adjust anything else.

Thanks! For parameterized tests, the idea is to write all 3 test cases as a parameterized single one. I do not think we need to parameterize on B, T, S etc.

**Comment by vfdev-5** [comment]
@ibbyml thanks for the updates! Please squash all commits into 1 otherwise CI will fail for num_commits >= 5

**Comment by vfdev-5** [comment]
@ibbyml let's keep only your updates

**Comment by ibbyml** [comment]
> @ibbyml let's keep only your updates

Sorry about that. Accidentally pulled everything. New push should be squashed correctly.

File: flax/nnx/nn/attention.py [hunk 3]
```diff
@@ -118,9 +124,17 @@ def dot_product_attention_weights(
...
     attn_weights = attn_weights + bias
   # apply attention mask
-  if mask is not None:
+  if mask is not None or is_causal:
     big_neg = jnp.finfo(dtype).min
-    attn_weights = jnp.where(mask, attn_weights, big_neg)
+    masks = [m for m in [mask] if m is not None]
+    if is_causal:
+      T, S = attn_weights.shape[-2:]
+      causal_mask = jnp.tril(jnp.ones((T, S), dtype=dtype))
```
Comment (vfdev-5) [inline flax/nnx/nn/attention.py:132]: @ibbyml sorry for the delay, we reviewed together with Cristian your PR. Can you please use [make_causal_mask](https://github.com/google/flax/blob/9ed59e2f9170d547facb81f15a2a0ccd8d786e60/flax/nnx/nn/attention.py#L767) to create causal mask instead of `jnp.tril`.
Thanks !

-> We can't use it as it produces a square causal matrix: `[..., S, S]` and we need in general a rectangular matrix: `[..., T, S]`
